### PR TITLE
CLOUDSTACK-9677: Adding storage policy support for swift as secondary…

### PR DESCRIPTION
**Pull request by khos2ow**
… storage

This is to bring PR #1830 back to life (since it was lost in time). We already have this fix in our prod well over a year now.

It's based on branch `4.11` and I guess it will be fine to go into `4.11.1` milestone.

### CODE REVIEW

**Comment by khos2ow** [comment]
@rhtyd @syed @rafaelweingartner Can you please take a look at this?

File: plugins/hypervisors/xenserver/src/com/cloud/hypervisor/xenserver/resource/XenServerStorageProcessor.java [hunk 1]
```diff
@@ -914,20 +917,45 @@ public Answer copyVolumeFromPrimaryToSecondary(final CopyCommand cmd) {
...
-            }
+            String result = hypervisorResource.callHostPluginAsync(conn, "swiftxenserver", "swift", wait, params.toArray(new String[params.size()]));
+            return "true".equals(result);
         } catch (final Exception e) {
             s_logger.warn("swift upload failed due to " + e.toString(), e);
         }
         return false;
     }
 
+    private List<String> getSwiftParams(SwiftTO swift, String container, String ldir, String lfilename, Boolean isISCSI) {
```
Comment (rafaelweingartner) [inline plugins/hypervisors/xenserver/src/com/cloud/hypervisor/xenserver/resource/XenServerStorageProcessor.java]: This method could really benefit from unit tests

File: plugins/hypervisors/xenserver/src/com/cloud/hypervisor/xenserver/resource/XenServerStorageProcessor.java [hunk 2]
```diff
@@ -914,20 +917,45 @@ public Answer copyVolumeFromPrimaryToSecondary(final CopyCommand cmd) {
...
+
         try {
-            result =
-                    hypervisorResource.callHostPluginAsync(conn, "swiftxenserver", "swift", wait, "op", "upload", "url", swift.getUrl(), "account", swift.getAccount(), "username",
-                            swift.getUserName(), "key", swift.getKey(), "container", container, "ldir", ldir, "lfilename", lfilename, "isISCSI", isISCSI.toString());
-            if (result != null && result.equals("true")) {
-                return true;
-            }
+            String result = hypervisorResource.callHostPluginAsync(conn, "swiftxenserver", "swift", wait, params.toArray(new String[params.size()]));
+            return "true".equals(result);
```
[the remark this answers is not in the document]
Comment (khos2ow) [reply plugins/hypervisors/xenserver/src/com/cloud/hypervisor/xenserver/resource/XenServerStorageProcessor.java]: Of course. +1.

**Comment by khos2ow** [comment]
@rafaelweingartner can you please review again?

File: plugins/hypervisors/xenserver/test/com/cloud/hypervisor/xenserver/resource/XenServerStorageProcessorTest.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        Assert.assertEquals("unexpected param.", "container", params.get(10));
+        Assert.assertEquals("unexpected param.", container, params.get(11));
+        Assert.assertEquals("unexpected param.", "ldir", params.get(12));
+        Assert.assertEquals("unexpected param.", ldir, params.get(13));
+        Assert.assertEquals("unexpected param.", "lfilename", params.get(14));
+        Assert.assertEquals("unexpected param.", lfilename, params.get(15));
+        Assert.assertEquals("unexpected param.", "isISCSI", params.get(16));
+        Assert.assertEquals("unexpected param.", isISCSI.toString(), params.get(17));
+
+        if (swift.getStoragePolicy() != null) {
```
Comment (rafaelweingartner) [inline plugins/hypervisors/xenserver/test/com/cloud/hypervisor/xenserver/resource/XenServerStorageProcessorTest.java]: @khos2ow thanks for creating the unit test case.
I have a doubt here. What is the execution flow you are testing? the one with storage policy or without it?
Also, what do you think about two distinct tests? One to execute the test with a storagePolicy and other without.

File: utils/src/main/java/com/cloud/utils/SwiftUtil.java [hunk 4]
```diff
@@ -294,12 +238,80 @@ public static String calculateRFC2104HMAC(String data, String key)
 
     }
 
-    public static String toHexString(byte[] bytes) {
+    static String toHexString(byte[] bytes) {
```
Comment (rafaelweingartner) [inline utils/src/main/java/com/cloud/utils/SwiftUtil.java]: What about using a commons to convert the string to Hexadecimal?
Something like: org.apache.commons.codec.binary.Hex.encode(byte[])

File: utils/src/test/java/com/cloud/utils/SwiftUtilTest.java [hunk 5]
```diff
@@ -90,4 +96,134 @@ public void testGetContainerName(){
 
         assertEquals(expected, output);
     }
+
+    @Test
```
Comment (rafaelweingartner) [inline utils/src/test/java/com/cloud/utils/SwiftUtilTest.java]: Unit Tests! That is awesome!!
Thank you very much for these. We need to create a culture of writing them.

File: utils/src/main/java/com/cloud/utils/SwiftUtil.java (hunk 4 above)
Comment (khos2ow) [reply utils/src/main/java/com/cloud/utils/SwiftUtil.java]: sure.

File: utils/src/test/java/com/cloud/utils/SwiftUtilTest.java (hunk 5 above)
Comment (khos2ow) [reply utils/src/test/java/com/cloud/utils/SwiftUtilTest.java]: I totally agree, but all the credit for this PS should go to @pdube. I only revived and rebased his PR.

File: plugins/hypervisors/xenserver/test/com/cloud/hypervisor/xenserver/resource/XenServerStorageProcessorTest.java (hunk 3 above)
Comment (khos2ow) [reply plugins/hypervisors/xenserver/test/com/cloud/hypervisor/xenserver/resource/XenServerStorageProcessorTest.java]: The flow is the basically test the exact order or params, and I actually do agree about two different tests for with and without storage policy.

**Comment by khos2ow** [comment]
@rhtyd @rafaelweingartner @wido @DaanHoogland Since 4.11 is out can we get these PR going again?

**Comment by rafaelweingartner** [comment]
I believe this one is still missing the automated testes to be executed.

**Comment by khos2ow** [comment]
@rafaelweingartner oh yes. In the meantime I'm gonna do the package until someone runs the test. @blueorangutan pacakge

**Comment by DaanHoogland** [comment]
unfortunatlelily blueorangutanisn't human and kan'd ried throe spelletjes
@blueorangutan package

**Comment by khos2ow** [comment]
damn! typo!!

**Comment by DaanHoogland** [comment]
not sure why mine didn't work, retrying
@blueorangutan package

**Comment by khos2ow** [comment]
@rhtyd @DaanHoogland do you have any idea why packaging is not finished yet? Is it normal?

**Comment by DaanHoogland** [comment]
@khos2ow no, it seems to never have started, retrying
@blueorangutan package

**Comment by khos2ow** [comment]
@DaanHoogland @rhtyd Can we get a round of test on this PR please?

**Comment by rhtyd** [comment]
@blueorangutan test centos7 xenserver-65sp1

**Comment by rhtyd** [comment]
Regression test LGTM, can you share your test results with swift @khos2ow and @pdube ?

**Comment by rafaelweingartner** [comment]
@khos2ow can you provide some inputs regarding @rhtyd inquiry? 
This one is ready to go.

**Comment by khos2ow** [comment]
Yes @rafaelweingartner , I will, soon.

**Comment by rhtyd** [comment]
Ping @khos2ow can you advise on testing of this PR against swift? This is ready to merge based on 2+lgtms and smoke test results.

**Comment by khos2ow** [comment]
Finally I managed some time to test this! (sorry for delay). I did couple round of manual tests as follow:
- created a secondary storage of type Swift with storage policy: `region-1`
- created a VM: `vm-1`
- created a volume (`vm-1-vol-1`) and attached to it
- created a snapshot of that volume and checked to see validated it ended up in `region-1`
- deleted secondary storage `region-1`
- created a secondary storage of type Swift with storage policy: `region-2`
- created a VM: `vm-2`
- created a volume (`vm-2-vol-1`) and attached to it
- created a snapshot of that volume and checked to see validated it ended up in `region-2`

**Comment by khos2ow** [comment]
@rhtyd @rafaelweingartner ping.

**Comment by rafaelweingartner** [comment]
Thanks for the replies @khos2ow. Is that what you wanted @rhtyd?

**Comment by rhtyd** [comment]
LGTM @khos2ow @rafaelweingartner. Since I don't any infra to test swift, I'll merge this based on reviews and previous automated regression tests and manual tests around swift shared by @khos2ow . Thanks for the effort.

**Comment by pdube** [comment]
Thanks @khos2ow

**Pull request merged**